packages/grafana-ui/src/types/panelOptions.tsx

```diff
@@ -0,0 +1,194 @@
+  optionType: OptionType;
+  /**
+   * Name of the options object property
+   * i.e. for GaugeOptions type:
+   * export const defaults: GaugeOptions = {
+   *   minValue: 0,
+   *   maxValue: 100,
+   *   ...
+   * }
+   * it could be 'minValue'
+   * Question: Should we name it option instead of path?
+   */
+  property: P;
+  /**
+   * Keeping component as optional on purpose to enable default components for simple types,
+   * i.e. number
+   */
+  component?: React.ComponentType<ComponentProps>;
+  label?: string;
+  placeholder?: string;
+}
+
+export interface OptionUIModel<TOptions, TKey extends keyof TOptions>
+  extends AbstractOptionUIModel<OptionUIComponentProps<TOptions[TKey]>, TKey> {}
+
+type LazySelectOptions<O> = () => Promise<O[]>;
+
+export interface SelectOptionUIModel<TOptions, TKey extends keyof TOptions>
+  extends AbstractOptionUIModel<SelectOptionUIComponentProps<TOptions[TKey]>, TKey> {
+  options: Array<SelectOptionItem<TOptions[TKey]>> | LazySelectOptions<SelectOptionItem<TOptions[TKey]>>;
+}
+
+export enum GroupLayoutType {
+  Panel = 'panel',
+  Fieldset = 'fieldset',
+}
+
+export interface GroupLayoutUIModel<K> {
+  type: GroupLayoutType;
+
+  /**
+   * Represents options of a given layout
+   */
+  groupOptions: K;
+
+  /**
+   * Represents panel options that belong to a given layout group
+   */
+  options: Array<OptionUIModel<any, any> | SelectOptionUIModel<any, any>>;
+}
+
+/**
+ * Represents a container for a group of options.
+ * Panel UI for options is characterised by a tit le with a gradient background
+ * Use it for grouping related options:
+ *
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      + Panel title                        +
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      + Option 1                    on/off +
+ *      + Option 2                    on/off +
+ *      + ...                                +
+ *      + ...                                +
+ *      ++++++++++++++++++++++++++++++++++++++
+ *
+ */
+export interface PanelUIModel extends GroupLayoutUIModel<{ title: string }> {}
+
+/**
+ * Represents a group of options
+ * Fieldset UI does not have any characteristic UI elements. It's basically a group of options
+ * with a title:
+ *
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      * Fieldset title                     *
+ *      *                                    *
+ *      + Option 1                    on/off +
+ *      + Option 2                    on/off +
+ *      + ...                                +
+ *      ++++++++++++++++++++++++++++++++++++++
+ */
+export interface FieldsetUIModel
+  extends GroupLayoutUIModel<{
+    legend: string;
+  }> {} // legend property name inspired by https://www.w3schools.com/tags/tag_legend.asp
+
+/**
+ * Below are typeguards used by PanelOptionsBuilder components
+ */
+export function isOptionsUIModel(model: any): model is OptionsUIModel<any> {
+  return (model as OptionsUIModel<any>).model !== undefined;
```

Comment:
what is the purpose of this dynamic type check, why not have model property on OptionsUIModel?

Reply:
`OptionsUIModel` has model property on it. This is a type guard to decide whether or not the automatic options builder should be used. See https://github.com/grafana/grafana/pull/16773/files#diff-6e9276e76a52b87735ffbbb4306fa5a8R71

There's probably a better way to decide if the panel uses declarative options or not, but I wanted to use `editor` prop on panel now to iterate fast